diluvian/volumes.py

```diff
@@ -1162,19 +1162,49 @@ def from_catmaid_stack(stack_info, tile_source_parameters):
         }[tile_source_parameters['tile_source_type']].format(**tile_source_parameters)
         bounds = np.flipud(np.array(stack_info['bounds'], dtype=np.int64))
         resolution = np.flipud(np.array(stack_info['resolution']))
+        translation = np.flipud(np.array(stack_info['translation']))
         tile_width = int(tile_source_parameters['tile_width'])
         tile_height = int(tile_source_parameters['tile_height'])
-        return ImageStackVolume(bounds, resolution, tile_width, tile_height, format_url,
-                                missing_z=stack_info['broken_slices'])
+        return ImageStackVolume(bounds, resolution, translation, tile_width, tile_height,
+                                format_url, missing_z=stack_info.get("broken_slices", None))
 
-    def __init__(self, bounds, orig_resolution, translation, tile_width, tile_height, 
+    def from_toml(filename):
+        volumes = {}
+        with open(filename, "rb") as fin:
+            datasets = toml.load(fin).get("ImageStack", [])
+            for dataset in datasets:
+                # stack info
+                si = [
+                    "bounds",
+                    "resolution",
+                    "translation",
+                    "broken_slices",
+                ]
+                # tile stack parameters
+                tsp = [
+                    "source_base_url",
+                    "file_extension",
+                    "tile_width",
+                    "tile_height",
+                    "tile_source_type",
+                ]
+                volume = ImageStackVolume.from_catmaid_stack(
+                    {si[key]: dataset[key] for key in si},
+                    {tsp[key]: dataset[key] for key in tsp},
+                )
+                volumes[dataset["title"]] = volume
+
+        return volumes
+
+    def __init__(self, bounds, orig_resolution, translation, tile_width, tile_height,
                  tile_format_url, zoom_level=0, missing_z=None, image_leaf_shape=None):
         self.orig_bounds = bounds
         self.orig_resolution = orig_resolution
         self.translation = translation
         self.tile_width = tile_width
         self.tile_height = tile_height
         self.tile_format_url = tile_format_url
+        self.mask_data = None
```

Comment:
This doesn't seem to be used. Is this fixing an existing bug with volumes assuming the `mask_data` attribute exists?

Reply:
The mask_data has to do with the PartitionedVolume class initializer on line 888 which requires its parent to have a mask_data field.

Reply:
👍